Migrate Competition Overview Page to React

FinnIckler opened the pull request:
Currently wip porting the registration requirements over

There is an issue that it only does one markdown request at once, but a bunch of fields are in markdown... so not sure what to do about that.

Progress:
![image](https://github.com/user-attachments/assets/a3589ca2-6ddf-4165-85c7-8d52294b5d1c)

FinnIckler commented:
Migrated all information over. Now onto the styling
<img width="1285" alt="Screenshot 2024-12-12 at 15 47 54" src="https://github.com/user-attachments/assets/3f9c3d8a-5fff-4a0d-afa1-9b244940773b" />
<img width="1285" alt="Screenshot 2024-12-12 at 15 48 03" src="https://github.com/user-attachments/assets/d5fc8330-8559-4677-9a2c-e51ed0d9f028" />

FinnIckler commented:
Moving it to the grid with the right sizes was quite easy actually. But I don't really know what the correct equivalent of a Descriptive List is in Semantic

FinnIckler commented:
I can rebuild it with another grid, maybe that looks good enough

FinnIckler commented:
<img width="655" alt="Screenshot 2024-12-12 at 21 42 04" src="https://github.com/user-attachments/assets/369c01c0-3bfa-440c-94aa-a788dde599a3" />
I was able to replicate it with a grid, but I did have to set the padding to 0 manually

FinnIckler commented:
<img width="1296" alt="Screenshot 2024-12-13 at 00 11 24" src="https://github.com/user-attachments/assets/6072c37d-e478-4444-9c61-64b7ed8398c8" />
<img width="1296" alt="Screenshot 2024-12-13 at 00 11 08" src="https://github.com/user-attachments/assets/b08a787e-7290-4a51-b1ba-b3424ce18905" />
Almost everything moved over to grid

gregorbg commented on app/webpacker/components/CompetitionTabs/Requirements/index.jsx:
Lots of dangling `br`, won't comment on all of them.
Not sure what you're trying to solve, but a simple `<List>` (which does _not_ have enumeration bullet points by default in SemUI!) might already do the trick in terms of line breaks.

FinnIckler replied:
managed to remove all those brs using a list

gregorbg commented on app/webpacker/components/CompetitionTabs/index.jsx:
```diff
@@ -0,0 +1,95 @@
+import './style.css';
+import TimeLimitCutoffInfo from './TimeLimitCutoffInfo';
+
+const updatePath = (tabSlug) => {
+  window.history.replaceState({}, '', `${window.location.pathname}#${tabSlug}`);
+};
```
Do we have shared `lib` functions for this? I feel like one or two other parts of our frontend code use similar features already

FinnIckler replied:
There is a function in filterUtils.js but it uses searchParams and not Hash so it works completely different

gregorbg commented on app/webpacker/components/CompetitionTabs/GeneralInfoTab.jsx:
Not happy about the repetitive `Grid`ding in here. Have you (a) tried out SemUI `List` using `List.Header`, `List.Description` and perhaps even `List.Icon` and (b) if a didn't work, can you please move the Grid cells into an "internal" component that you can reuse at least within this page?

FinnIckler replied:
It does not work with a List, but I did do a bunch of refactoring to reuse the grid components

FinnIckler commented on app/webpacker/components/CompetitionTabs/Requirements/RegistrationFeeRequirements.jsx:
```diff
@@ -0,0 +1,27 @@
+          })
+          : I18n.t('competitions.competition_info.no_entry_fee')}
+      </p>
+      {competition.competition_events.map((event) => (event['has_fee?'] ? (
+        <dl key={event.id}>
+          <dt>{event.event.name}</dt>
+          <dd>{isoMoneyToHumanReadable(event.fee)}</dd>
+        </dl>
+      ) : null))}
```
this is just not migrated yet!

FinnIckler replied:
we don't have any way to currently set event based registration fees right? So can this not be safely removed?

FinnIckler commented on app/webpacker/components/CompetitionTabs/Requirements/GuestRequirements.jsx:
```diff
@@ -0,0 +1,31 @@
+              competition.currency_code,
+            ),
+        })}
+        <br />
```
the <br /> are from the original file

FinnIckler commented:
![image](https://github.com/user-attachments/assets/1974360d-0737-4ee1-bb11-e5a58b7da768)
Mobile view

kr-matthews commented on app/webpacker/components/BookmarkIcon/index.jsx:
```diff
@@ -0,0 +1,49 @@
+  const { mutate: bookmarkCompetitionMutation, isPending: isBookmarking } = useMutation({
+    mutationFn: bookmarkCompetition,
+    onSuccess: () => {
+      setIsBookmarked(!isBookmarked);
```
It's safer to explicitly set it, just in case the state somehow gets out of sync. It might be possible for the user to click again while the initial mutation is processing or something, which might flip it twice? (Might be worth disabling the icon while it's loading?)
```suggestion
      setIsBookmarked(true);
```
Likewise below, but with false.

Also, could be simplified to one line:
```
onSuccess: () => setIsBookmarked(true),
```

kr-matthews commented on app/webpacker/components/CompetitionTabs/Requirements/GuestRequirements.jsx:
This seems like a mix of an if and a ternary, where the second case of the ternary is also conditional. It might be clearer as 3 if statements.

kr-matthews reviewed (commented):
(Only viewed some of the files. Will do the rest later, if the PR is still open.)

gregorbg commented:
Opinions?
![image](https://github.com/user-attachments/assets/89f68c2f-4430-434d-bc5c-6b45e3fd62d6)

kr-matthews reviewed (changes requested):
~I haven't fully reviewed everything yet.~

I'd be in favour of having the main, exported function at the top of files.

kr-matthews commented on app/webpacker/components/CompetitionTabs/WinnerTable.jsx:
```diff
@@ -0,0 +1,81 @@
+            return (
+              <Table.Row>
+                <Table.Cell>
+                  <a href={competitionAllResultsUrl(competition.id, r.event.id)}>
```
I'm getting an error when I try to view results for a particular event.

FinnIckler commented:
> ~I haven't fully reviewed everything yet.~
> 
> I'd be in favour of having the main, exported function at the top of files.

Will that not cause a 'referenced before defined' eslint error?

FinnIckler commented:
Moved the numbers on the right to a Statisticgroup
![image](https://github.com/user-attachments/assets/40f36dad-4387-46a0-b89f-e8457b67c673)

Also added a 'Registrations' statistic that is shown instead of the competitor count when the competition isn't finished yet

FinnIckler commented on app/webpacker/components/CompetitionTabs/GeneralInfo/GeneralInfoTab.jsx:
```diff
@@ -0,0 +1,136 @@
+import DateAddressContact from './DateAddressContact';
+import InformationEvents from './InformationEvents';
+
+function RegistrationTime({ competition }) {
```
I removed the seconds. About the GMT-8 vs. PST, I don't know why that happens for you, but https://github.com/moment/luxon/blob/master/docs/formatting.md shows that DATETIME_FULL should use the offset name and not just the offset
